### PR TITLE
Add documentation for close()

### LINKED ISSUE #1010: Public documentation is missing for close()
* Description of issue

Documentation on readthedocs is missing an entry for `close()`

### PR DESCRIPTION
**Pull request by texasaggie97**
- [X] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/nimi-python/blob/master/CONTRIBUTING.md).
- [X] ~~I've updated [CHANGELOG.md](https://github.com/ni/nimi-python/blob/master/CHANGELOG.md) if applicable.~~
- [X] I've added tests applicable for this pull request
### What does this Pull Request accomplish?
* Add documentation for close()
* Updated .codecov.yml for current state
### List issues fixed by this Pull Request below, if any.
* Fix #1010 

### What testing has been done?
* Visual inspection of generated files

### CODE REVIEW

File: build/helper/documentation_snippets.py [hunk 1]
```diff
@@ -96,5 +96,97 @@
 { 'simulate': True, 'driver_setup': { 'Model': '<model number>',  'BoardType': '<type>' } }
 '''
 
+default_close_function_doc = '''
+Closes the driver session and cleans up.
+'''
+
+close_function_note = '''
+This function is not needed when using the session context manager
```
Comment (marcoskirsch) [inline build/helper/documentation_snippets.py]: Do you think clients will understand what "the session context manager" is?
Honest question.

Comment (texasaggie97) [reply build/helper/documentation_snippets.py]: By default we will use the documentation from the driver's close function if it exists. 

The note will be added regardless of whether the default description is used or the description from the driver. I am open on what the note should say to make clear that normally you don't need to use this function. I agree this isn't very good, but couldn't come up with a better way to say this.

Comment (marcoskirsch) [reply build/helper/documentation_snippets.py]: Looking at docs:
https://docs.python.org/3/whatsnew/2.6.html#pep-343-the-with-statement

They talk of the `with` statement, rather than of context managers.
I suggest following this terminology here as well as in initate/abort/initalize.

Let's refine that in its own PR so we don't hold this one.

File: build/helper/documentation_snippets.py [hunk 2]
```diff
@@ -96,5 +96,97 @@
...
+'''
+
+def close_function_def_for_doc(functions):
+    # This is very specific to IVI. We look for a 'close' function and if we find it,
```
Comment (marcoskirsch) [inline build/helper/documentation_snippets.py]: NISE and niModInst have close. How is this specific to IVI?

File: docs/nidigital/class.rst [hunk 3]
```diff
@@ -3685,6 +3687,21 @@ clock_generator_initiate
...
+
+    .. py:currentmodule:: nidigital.Session
+
+    .. py:method:: close()
+
+            Closes the driver session and cleans up.
+
+            
+
+            .. note:: This method is not needed when using the session context manager
```
Comment (marcoskirsch) [inline docs/nidigital/class.rst]: We shall refine the wording.

File: generated/nidcpower/session.py [hunk 4]
```diff
@@ -3724,6 +3724,26 @@ def initiate(self):
...
+        that NI-DCPower reserves. If power output is enabled when you call this
+        method, the output channels remain in their existing state and
+        continue providing power. Use the ConfigureOutputEnabled
+        method to disable power output on a per channel basis. Use the
+        reset method to disable power output on all channel(s).
+
+        **Related Topics:**
+
+        `Programming
+        States <REPLACE_DRIVER_SPECIFIC_URL_1(programmingstates)>`__
```
Comment (marcoskirsch) [inline generated/nidcpower/session.py]: ?

File: build/helper/documentation_snippets.py (hunk 1 above)
Comment (marcoskirsch) [reply build/helper/documentation_snippets.py]: https://github.com/ni/nimi-python/issues/1020

File: build/helper/documentation_snippets.py (hunk 2 above)
Comment (texasaggie97) [reply build/helper/documentation_snippets.py]: Changed to `# This is very specific to session based APIs ...`

File: generated/nidcpower/session.py (hunk 4 above)
Comment (texasaggie97) [reply generated/nidcpower/session.py]: This is to make fixing up links in help. The original link is internal to the help file and does not work here, so I replace it with something that can be easily searched for.

This is not new, pre 1.0.

File: build/helper/documentation_snippets.py [hunk 5]
```diff
@@ -96,6 +96,14 @@
 { 'simulate': True, 'driver_setup': { 'Model': '<model number>',  'BoardType': '<type>' } }
 '''
 
+default_close_function_doc = '''
+Closes the driver session and cleans up.
```
Comment (marcoskirsch) [inline build/helper/documentation_snippets.py]: Should we add: "After calling this, the Session object is no longer valid and cannot be used."

File: build/helper/documentation_snippets.py [hunk 6]
```diff
@@ -96,6 +96,14 @@
...
+'''
+
+close_function_note = '''
+This function is not needed when using the session context manager
+'''
+
 default_initiate_function_doc = '''
 Calls initiate
```
Comment (marcoskirsch) [inline build/helper/documentation_snippets.py]: > "`Calls initiate`"
Is this the documentation we'd get for `initiate()`?
It's useless, no?

File: build/helper/documentation_snippets.py [hunk 7]
```diff
@@ -105,6 +113,45 @@
...
+            function_def['documentation'] = {}
+        if 'description' not in function_def['documentation']:
+            function_def['documentation']['description'] = default_close_function_doc
+        if 'note' not in function_def['documentation']:
+            function_def['documentation']['note'] = []
+        if type(function_def['documentation']['note']) is not list:
+            function_def['documentation']['note'] = [function_def['documentation']['note']]
+        function_def['documentation']['note'].append(close_function_note)
+        function_def['python_name'] = 'close'
+    else:
```
Comment (marcoskirsch) [inline build/helper/documentation_snippets.py]: Maybe we should consider breaking the build if we don't have documentation for `close` (or `initiate`).
I assume having the `else` case is nice right now because `nidigital` doesn't have documentation. If that's the reason for having this, we should consider adding documentation via add-on just for these two functions.

File: build/helper/documentation_snippets.py (hunk 5 above)
Comment (texasaggie97) [reply build/helper/documentation_snippets.py]: Yes, that seems like a good thing to add.

File: build/helper/documentation_snippets.py (hunk 6 above)
Comment (texasaggie97) [reply build/helper/documentation_snippets.py]: This is only used if initiate doesn't have its own documentation. In theory, you should never see this. Given that, I am open to a better, generic description.

Comment (marcoskirsch) [reply build/helper/documentation_snippets.py]: > This is only used if initiate doesn't have its own documentation.

I'd rather have this be a build error.

File: build/helper/documentation_snippets.py (hunk 7 above)
Comment (texasaggie97) [reply build/helper/documentation_snippets.py]: Done. Assert when close or initiate function doesn't exist

Also found that `'close_function'` in config metadata was incorrect. Fixed here and internally.

File: .codecov.yml [hunk 8]
```diff
@@ -6,9 +6,10 @@ coverage:
   status:
     project:
       default:
-        threshhold: 5%
+        target: 90%
```
Comment (marcoskirsch) [inline .codecov.yml]: This change appears to be unrelated to the PR, as it is not mentioned in the description.
Is this a merge problem?

File: build/helper/documentation_snippets.py [hunk 9]
```diff
@@ -105,6 +114,29 @@
 '''
 
 
+def close_function_def_for_doc(functions, config):
+    # This is very specific to session based APIs. We look for a 'close' function and if we find it,
```
Comment (marcoskirsch) [inline build/helper/documentation_snippets.py]: I think this is missing context and is now a bit outdated. Suggestion:


> This is very specific to session based APIs. We look for a 'close' function in the metadata which is to become a code-generated "private" method of Session for which we provide a public wrapper.
Copy its documentation so we apply it to the said public wrapper.

File: build/helper/documentation_snippets.py [hunk 10]
```diff
@@ -96,6 +96,15 @@
 { 'simulate': True, 'driver_setup': { 'Model': '<model number>',  'BoardType': '<type>' } }
 '''
 
+default_close_function_doc = '''
+Closes the driver session and cleans up. After calling this the Session object
+is no lonber valid and cannot be used.
```
Comment (marcoskirsch) [inline build/helper/documentation_snippets.py]: longer (typo)

File: build/helper/documentation_snippets.py [hunk 11]
```diff
@@ -105,6 +114,29 @@
...
+        function_def['python_name'] = 'close'
+    else:
+        assert False, "No '{}' function defined".format(close_name)
+
+    return function_def
+
+
 def initiate_function_def_for_doc(functions, config):
     # This is very specific to the IVI state model which not all drivers in nimi-python follow.
     # We look for a 'close' function and if we find it, We will copy that and modify it to be
```
Comment (marcoskirsch) [inline build/helper/documentation_snippets.py]: this is also outdated... there is no more "if we find it".

File: .codecov.yml (hunk 8 above)
Comment (texasaggie97) [reply .codecov.yml]: It is the same change as is in #1023 to correctly control the limits.

I will add something to the description to mention it.

File: build/helper/documentation_snippets.py (hunk 6 above)
Comment (texasaggie97) [reply build/helper/documentation_snippets.py]: You want to require there be supplied documentation? Is that reasonable?

File: build/helper/documentation_snippets.py (hunk 9 above)
Comment (texasaggie97) [reply build/helper/documentation_snippets.py]: Changed

File: build/helper/documentation_snippets.py (hunk 10 above)
Comment (texasaggie97) [reply build/helper/documentation_snippets.py]: Fixed

File: build/helper/documentation_snippets.py (hunk 11 above)
Comment (texasaggie97) [reply build/helper/documentation_snippets.py]: Changed to something similar as for close.
